File: Algorithm/C++/1000+/1034. Coloring A Border.cc
/*
 * Created on Sun May 05 2019 22:26:48
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// bfs, record borders of connected component
#include <vector>
#include <queue>

using std::vector;
using std::queue;

class Solution {
 public:
  vector<vector<int>> colorBorder(vector<vector<int>>& grid, int r0, int c0, int color) {
    int m = grid.size(), n = grid[0].size();
    vector<vector<int>> visited(m, vector<int>(n, 0));
    vector<vector<int>> borders;
    queue<vector<int>> q;
    q.push(vector<int>{r0, c0});
    visited[r0][c0] = 1;
    vector<vector<int>> directions {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    while (q.size() > 0) {
      vector<int> curr = q.front();
      q.pop();
      for (auto d : directions) {
        if (curr[0] == 0 || curr[0] == m - 1 || curr[1] == 0 || curr[1] == n - 1)
          borders.push_back(vector<int>{curr[0], curr[1]});
        int r = curr[0] + d[0], c = curr[1] + d[1];
        if (r >= 0 && r < m && c >= 0 && c < n) {
          if (grid[r][c] == grid[r0][c0]) {
            if (visited[r][c] == 0)
              q.push(vector<int>{r, c});
          } else {
            borders.push_back(vector<int>{curr[0], curr[1]});
          }
          visited[r][c] = 1;
        }
      }
    }
    for (auto b : borders) {
      grid[b[0]][b[1]] = color;
    }
    return grid;
  }
};

File: Algorithm/Python/1000+/1034. Coloring A Border.py
# -*- coding: utf-8 -*-
# Created on Sun May 05 2019 10:27:22
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# bfs, record borders of connected component

from collections import deque

class Solution(object):
    def colorBorder(self, grid, r0, c0, color):
        """
        :type grid: List[List[int]]
        :type r0: int
        :type c0: int
        :type color: int
        :rtype: List[List[int]]
        """
        queue, borders = deque(), []
        m, n = len(grid), len(grid[0])
        visited = [[0] * n for _ in xrange(m)]
        queue.append((r0, c0))
        visited[r0][c0] = 1
        directions = ((-1, 0), (1, 0), (0, -1), (0, 1))
        while queue:
            r, c = queue.popleft()
            if r == 0 or r == m - 1 or c == 0 or c == n - 1:
                borders.append((r, c))
            for d in directions:
                nr, nc = r + d[0], c + d[1]
                if 0 <= nr < m and  0 <= nc < n:
                    if grid[nr][nc] == grid[r0][c0]:
                        if visited[nr][nc] == 0:
                            queue.append((nr, nc))
                    else:
                        borders.append((r, c))
                    visited[nr][nc] = 1
        for r, c in borders:
            grid[r][c] = color
        return grid
    